packages/client/test/events.spec.ts

```diff
@@ -2,198 +2,235 @@ import {
-} as unknown as Provider;
-
-const MOCK_ERROR_EVENT_PROVIDER = {
-  metadata: {
-    name: 'mock-events-failure',
-  },
-  initialize: jest.fn(() => {
-    return Promise.reject({
-      reason: ERROR_REASON,
-      errorCode: ERROR_CODE,
-    });
-  }),
-  events: new OpenFeatureEventEmitter(),
-} as unknown as Provider;
-
-const MOCK_NO_EVENT_PROVIDER = {
-  metadata: {
-    name: 'mock-no-events',
-  },
-  initialize: jest.fn(() => {
-    return Promise.resolve(undefined);
-  }),
-  // no events.
-} as unknown as Provider;
-
-const MOCK_ERROR_NO_EVENT_PROVIDER = {
-  metadata: {
-    name: 'mock-no-events-failure',
-  },
-  initialize: jest.fn(() => {
-    return Promise.reject({
-      reason: ERROR_REASON,
-      errorCode: ERROR_CODE,
-    });
-  }),
-  // no events.
-} as unknown as Provider;
+class MockProvider implements Provider {
+  readonly metadata: ProviderMetadata;
+  readonly events?: OpenFeatureEventEmitter;
+  private hasInitialize: boolean;
+  private failOnInit: boolean;
+  private enableEvents: boolean;
+  status?: ProviderStatus = undefined;
+
+  constructor(options?: {
+    hasInitialize?: boolean;
+    initialStatus?: ProviderStatus;
+    enableEvents?: boolean;
+    failOnInit?: boolean;
+    name?: string;
+  }) {
+    this.metadata = { name: options?.name ?? 'mock-provider' };
+    this.hasInitialize = options?.hasInitialize ?? true;
+    this.status = options?.initialStatus ?? ProviderStatus.NOT_READY;
+    this.enableEvents = options?.enableEvents ?? true;
+    this.failOnInit = options?.failOnInit ?? false;
+
+    if (this.enableEvents) {
+      this.events = new OpenFeatureEventEmitter();
+    }
+
+    if (this.hasInitialize) {
+      this.initialize = jest.fn(async () => {
+        if (this.failOnInit) {
+          throw {
+            reason: ERROR_REASON,
+            errorCode: ERROR_CODE,
+          };
+        }
+
+        this.status = ProviderStatus.READY;
+      });
+    }
+  }
+
+  initialize: jest.Mock<Promise<void>, []> | undefined;
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  async onClose(): Promise<void> {}
+
+  resolveBooleanEvaluation(): ResolutionDetails<boolean> {
+    throw new Error('Not implemented');
+  }
+
+  resolveNumberEvaluation(): ResolutionDetails<number> {
+    throw new Error('Not implemented');
+  }
+
+  resolveObjectEvaluation<T extends JsonValue>(): ResolutionDetails<T> {
+    throw new Error('Not implemented');
+  }
+
+  resolveStringEvaluation(): ResolutionDetails<string> {
+    throw new Error('Not implemented');
+  }
+}
 
 describe('Events', () => {
   // set timeouts short for this suite.
   jest.setTimeout(1000);
-  let API: OpenFeatureAPI = new (OpenFeatureAPI as any)();
+  let clientId = uuid();
 
   afterEach(() => {
-    jest.clearAllMocks();
-    API = new (OpenFeatureAPI as any)(); //TODO this hacky thing should be removed with provider mapping: https://github.com/open-feature/js-sdk/issues/421
+    // jest.clearAllMocks();
```

Comment:
Should this be commented?

Reply:
@toddbaert I changed that now.
The problem was, that there was actually a problem with the implementation of the events.
Events that came up in one provider have been emitted to every client.
I changed that a bit so that we have global and client events.
I would continue on that in my PR for events 🙂

Reply:
Yes, when we implement events in server we should also add tests for this specific use-case (check that provider x doesn't cause events for a client associated with provider y).

Since events are only implemented experimentally in web, I'm OK with doing that in a later PR.